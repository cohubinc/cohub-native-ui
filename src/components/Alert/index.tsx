import React, { ReactNode } from "react";
import { View, ViewProps } from "react-native";
import { Color } from "@cohubinc/cohub-utils";

import Typography from "src/components/Typography";

export interface IAlertProps extends ViewProps {
  traceProp?: any;
  centerAlign?: boolean;
  info?: boolean;
  success?: boolean;
  error?: boolean;
  message?: ReactNode | string[];
}

export function Alert(props: IAlertProps) {
  const {
    style,
    error,
    message,
    traceProp,
    success,
    info,
    centerAlign,
    ...rest
  } = props;

  let backgroundColor: any = "#f8f8f9";
  let color: any = "rgba(0,0,0,.87)";

  if (error) {
    color = Color.red800;
    backgroundColor = Color.red100;
  } else if (success) {
    color = Color.green900;
    backgroundColor = Color.green100;
  } else if (info) {
    color = Color.blue800;
    backgroundColor = Color.blue100;
  }

  const AlertText = ({ children }: { children: string }) => (
    <Typography.Small color={color}>{children}</Typography.Small>
  );
  function renderChildren() {
    if (typeof message === "string") {
      return <AlertText>{message}</AlertText>;
    }

    if (
      Array.isArray(message) &&
      message.every(msg => typeof msg === "string")
    ) {
      return (message as string[]).map(text => {
        return <AlertText key={text}>{text}</AlertText>;
      });
    }

    return message;
  }

  return (
    <View
      style={[
        {
          flexDirection: centerAlign ? "row" : "column",
          justifyContent: "center",
          alignItems: "center",
          paddingVertical: 16,
          paddingHorizontal: 20,
          borderRadius: 4,
          backgroundColor
        },
        style
      ]}
      {...rest}
    >
      <View>{renderChildren()}</View>
    </View>
  );
}
