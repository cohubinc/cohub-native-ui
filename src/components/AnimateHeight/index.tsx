import React, { SFC, useState, useRef, useEffect, Children } from "react";
import {
  ViewStyle,
  StyleProp,
  Animated,
  LayoutChangeEvent,
  View
} from "react-native";

interface IAnimateHeightProps {
  expanded: boolean;
  style?: StyleProp<ViewStyle>;
}
const AnimateHeight: SFC<IAnimateHeightProps> = props => {
  const { children, expanded, style } = props;

  const [elHeight, setElHeight] = useState<number | undefined>(undefined);

  const expandAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (elHeight === undefined) return;

    Animated.timing(expandAnimation, {
      toValue: expanded ? elHeight : 0
    }).start();
  }, [expanded]);

  // This is just to get the UI in the correct state after we are able to calculate the element height
  // It should only run once, just after elHeight is set the first time
  useEffect(() => {
    if (elHeight === undefined) return;

    expandAnimation.setValue(expanded ? elHeight : 0);
  }, [elHeight === undefined]);

  function getElHeightOnLayout({ nativeEvent }: LayoutChangeEvent) {
    // Only get set the height once
    if (elHeight !== undefined) return;

    setElHeight(nativeEvent.layout.height);
  }
  if (!Children.count(children)) {
    return null;
  }

  interface IStyle {
    opacity?: number;
    height?: number | Animated.Value;
    overflow: string;
  }
  let derivedStyle: IStyle = { height: expandAnimation, overflow: "hidden" };
  if (elHeight === undefined) {
    // If element height isn't calculated yet set opacity to 0 so we can get a measurement but without actually revealing the element
    derivedStyle.opacity = 0;
    // If element height isn't calculated yet don't set the height so we can measure it's 'natural' height first
    derivedStyle.height = undefined;
  }

  return (
    <View style={style}>
      <Animated.View style={[derivedStyle]} onLayout={getElHeightOnLayout}>
        {children}
      </Animated.View>
    </View>
  );
};

export default AnimateHeight;
