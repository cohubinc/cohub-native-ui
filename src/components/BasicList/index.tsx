import React, { useState } from "react";
import {
  FlatList,
  FlatListProps,
  ListRenderItem,
  View,
  ListRenderItemInfo,
  RefreshControl
} from "react-native";
import { Color } from "@cohubinc/cohub-utils";

import Loader from "src/components/Loader";
import Divider from "src/components/Divider";

interface ILoaderRow {
  isLoaderRow?: boolean;
}
type IOnEndReachedCallback<Resp = void> = (info: {
  distanceFromEnd: number;
}) => Resp;
type IFlatListProps<TItem> = Omit<
  FlatListProps<TItem | ILoaderRow>,
  "data" | "renderItem" | "onEndReached" | "refreshControl"
>;
export interface IBasicListProps<TItem> extends IFlatListProps<TItem> {
  loading?: boolean;
  data?: Array<TItem | ILoaderRow>;
  renderItem: ListRenderItem<TItem>;
  onEndReached?:
    | IOnEndReachedCallback
    | IOnEndReachedCallback<Promise<() => void>>;
}
export default function BasicList<TItem>(props: IBasicListProps<TItem>) {
  let {
    keyExtractor = defaultKeyExtractor,
    data = [],
    renderItem,
    onEndReached,
    loading,
    onRefresh,
    ItemSeparatorComponent,
    style,
    ...rest
  } = props;

  const [fetchingMore, setFetchingMore] = useState(false);

  if (data.length && fetchingMore) {
    data = [...data, { isLoaderRow: true }];
  }
  const isLoading = loading || fetchingMore;

  return (
    <FlatList
      {...{ onRefresh, data, keyExtractor, ...rest }}
      style={[style]}
      renderItem={rowData => {
        const { item } = rowData;
        if ("isLoaderRow" in item && item.isLoaderRow) {
          return <LoadingIndicatorRow />;
        }

        return renderItem(rowData as ListRenderItemInfo<TItem>);
      }}
      refreshControl={
        <RefreshControl
          tintColor={Color.green300 as any}
          refreshing={isLoading}
          onRefresh={onRefresh || undefined}
        />
      }
      ItemSeparatorComponent={() => <Divider />}
      refreshing={isLoading}
      onEndReachedThreshold={0.5}
      onEndReached={async info => {
        if (!onEndReached) return;
        setFetchingMore(true);
        await onEndReached(info);
        setFetchingMore(false);
      }}
    />
  );
}

const LoadingIndicatorRow = () => (
  <View
    style={[
      {
        justifyContent: "center",
        alignItems: "center",
        height: 30,
        marginBottom: 15
      }
    ]}
  >
    <Loader size={15} />
  </View>
);

type IRowItem<TItem> = {
  id?: string | number;
} & TItem;

function defaultKeyExtractor<TItem>(item: IRowItem<TItem>, index: number) {
  return (item && item.id ? item.id : index).toString();
}
