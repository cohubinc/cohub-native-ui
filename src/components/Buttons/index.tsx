import React, { Component } from "react";

import Button from "./Base";
import Blank from "./Blank";
import OutlineButton from "./Outline";
import { Ghost, PrimaryGhostButton, CancelGhostButton } from "./Ghost";
import Text from "./Text";
import IButtonProps from "../../definitions/interfaces/IButtonProps";

import Primary from "./Primary";
// import Secondary from "./Secondary";
// import Info from "./Info";
// import Cancel from "./Cancel";
// import Dropdown from "./Dropdown";
// import FloatingActionButton from "./FloatingActionButton";
import Split from "./Split";

// export { TBlankButtonProps } from "./Blank";
// export { TPrimaryButtonProps } from "./Primary";
// export { TOutlineButtonProps } from "./Outline";
// export { TGhostButtonProps } from "./Ghost";
// export { TTextButtonProps } from "./Text";
// export { IDropdownButtonProps } from "./Dropdown";
// export { TFloatingActionButtonProps } from "./FloatingActionButton";
// export { ISplitButtonProps } from "./Split";

export default class Buttons extends Component<IButtonProps> {
  static Base = Button;
  static Primary = Primary;
  // static Secondary = Secondary;
  // static Info = Info;
  // static Cancel = Cancel;
  static Outline = OutlineButton;
  // static Ghost = Ghost;
  // static PrimaryGhost = PrimaryGhostButton;
  // static CancelGhost = CancelGhostButton;

  // static Dropdown = Dropdown;

  // static Text = Text;

  // static Blank = Blank;

  // static FloatingAction = FloatingActionButton;

  static Split = Split;

  render() {
    return <Buttons.Primary {...this.props} />;
  }
}
