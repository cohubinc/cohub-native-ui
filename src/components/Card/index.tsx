import Horizontal from "./Horizontal";
import Vertical from "./Vertical";

const Card = {
  Horizontal,
  Vertical
};

export default Card;
