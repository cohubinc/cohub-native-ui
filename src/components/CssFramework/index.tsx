import React from "react";
import kebabCase from "lodash/kebabCase";

import Color from "src/definitions/enums/Color";
import BoxShadow from "src/definitions/enums/BoxShadow";
import Font from "src/definitions/enums/Font";

import "./styles/main.scss";

enum Miscellaneous {
  defaultBorderRadius = "4px",
  disabledOpacity = "0.45",
  gutter = "16px"
}

const variablesString = `
  :root {
    ${generateCssVariables(BoxShadow)}
    ${generateCssVariables(Color)}
    ${generateCssVariables(Font)}
    ${generateCssVariables(Miscellaneous)}
  }
`;

const CssVariables = () => <style>{variablesString}</style>;

type CSSENumType =
  | typeof Color
  | typeof Font
  | typeof Miscellaneous
  | typeof BoxShadow;

function generateCssVariables(VariablesEnum: CSSENumType) {
  const names = Object.keys(VariablesEnum).filter(name => name.indexOf("hsl("));

  const cssVariables = names.reduce((accumulatedStyles, name) => {
    const variableName = kebabCase(name);
    const value: string = (VariablesEnum as any)[name];

    return `${accumulatedStyles}
      --${variableName}: ${value};
    `;
  }, "");

  return cssVariables;
}
// generateCssVariables will return a string that looks something like whats below
/*
  --cohub-green: #63B05A;

  --black:  #222020;

  --dark-black: #000;
*/

export default CssVariables;
