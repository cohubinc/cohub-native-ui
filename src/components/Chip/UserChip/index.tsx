import React from "react";
import Chip from "..";
import { IUser } from "src/components/graphql";
import Avatar from "src/components/Avatar";
import Typography from "src/components/Typography";
import { TProps as ChipProps } from "..";

interface IProps extends ChipProps {
  user: IUser;
}

export default class UserChip extends React.PureComponent<IProps> {
  render() {
    const { user, ...rest } = this.props;

    return (
      <Chip
        key={user.id}
        style={{
          paddingTop: "4px",
          paddingBottom: "4px",
          marginRight: "4px"
        }}
        dark
        className="mb-1"
        {...rest}
      >
        <div className="flex items-center">
          <Avatar
            size={20}
            src={user.avatar && user.avatar.small_url}
            className="mr-05"
          />
          <Typography.Small>{user.name}</Typography.Small>
        </div>
      </Chip>
    );
  }
}
