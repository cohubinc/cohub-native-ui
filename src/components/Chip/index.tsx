import React, { MouseEventHandler, Component } from "react";

import Color, { ContrastColor } from "src/definitions/enums/Color";
import Typography from "src/components/Typography";
import Icon, { IIconProps } from "src/components/Icon";
import HtmlElementProps from "src/definitions/types/HtmlElementProps";

import AddChipInput from "./AddChipInput";
import AvatarChip from "./AvatarChip";

import "./Chip.scss";

interface IChipProps {
  label?: string;
  onDelete?: MouseEventHandler<HTMLElement>;
  checked?: boolean;
  dark?: boolean;
  backgroundColor?: Color;
  size?: number;
  active?: boolean;
}

export type TChipProps = IChipProps &
  Pick<HtmlElementProps, "onClick" | "className" | "style">;

export default class Chip extends Component<TChipProps> {
  static Add = AddChipInput;
  static Avatar = AvatarChip;

  static defaultProps: Partial<TChipProps> = {
    size: 12,
    backgroundColor: Color.grey300,
    active: false
  };

  render() {
    const {
      children,
      label,
      onClick,
      onDelete,
      checked,
      className = "",
      backgroundColor,
      style,
      size,
      active
    } = this.props;

    const name = label || children;
    const clickable = !!onClick;

    const clickableClass = clickable ? "clickable" : "";

    let iconName: IIconProps["name"] | undefined;
    if (checked) {
      iconName = "checkmark";
    } else if (onDelete) {
      iconName = "close";
    }

    const padding = `${size! / 2.5}px ${size}px`;

    const setBackgroundColor = () => {
      if (active) {
        return Color.green500;
      } else {
        return backgroundColor as Color;
      }
    };

    return (
      <div
        className={`CohubChip ${clickableClass} ${className}`}
        style={{
          backgroundColor: setBackgroundColor() as any,
          borderRadius: "361px",
          display: "inline-block",
          padding,
          ...style
        }}
        onClick={onClick}
        tabIndex={clickable ? 0 : undefined}
      >
        <div
          className="flex justify-center items-center h-100"
          style={{ cursor: clickable ? "pointer" : "inherit" }}
        >
          <Typography.Small color={ContrastColor[setBackgroundColor()] as any}>
            {name}
          </Typography.Small>

          {iconName && (
            <Icon
              onPress={e => onDelete && onDelete(e as any)}
              size={16}
              name={iconName}
              color={ContrastColor[setBackgroundColor()] as any}
            />
          )}
        </div>
      </div>
    );
  }
}
