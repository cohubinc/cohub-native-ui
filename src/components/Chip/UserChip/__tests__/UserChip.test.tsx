import React from "react";
import { mount } from "enzyme";

import UserChip from "../index";
import { IUser } from "src/components/graphql";

it("renders without crashing", async () => {
  const wrapper = mount(
    <UserChip
      user={
        { id: "1", avatar: { small_url: "test" }, name: "Test Guy" } as IUser
      }
    />
  );

  expect(wrapper).toMatchSnapshot();
});
