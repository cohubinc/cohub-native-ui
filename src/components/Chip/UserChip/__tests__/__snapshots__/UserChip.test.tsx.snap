// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<UserChip
  user={
    Object {
      "avatar": Object {
        "small_url": "test",
      },
      "id": "1",
      "name": "Test Guy",
    }
  }
>
  <Chip
    backgroundColor="hsl(0, 0%, 96%)"
    className="mb-1"
    dark={true}
    key="1"
    size={12}
    style={
      Object {
        "marginRight": "4px",
        "paddingBottom": "4px",
        "paddingTop": "4px",
      }
    }
  >
    <div
      className="CohubChip  mb-1"
      style={
        Object {
          "backgroundColor": "hsl(0, 0%, 96%)",
          "borderRadius": "361px",
          "display": "inline-block",
          "marginRight": "4px",
          "padding": "6px 12px",
          "paddingBottom": "4px",
          "paddingTop": "4px",
        }
      }
    >
      <div
        className="flex justify-center items-center h-100"
        style={
          Object {
            "cursor": "inherit",
          }
        }
      >
        <Component>
          <span
            className=""
            data-qa="text"
            style={
              Object {
                "color": "hsl(0, 2%, 22%)",
                "fontFamily": "Inter",
                "fontSize": 12,
                "fontWeight": 400,
                "textAlign": "initial",
                "textTransform": "initial",
              }
            }
          >
            <div
              className="flex items-center"
            >
              <Avatar
                className="mr-05"
                size={20}
                src="test"
              >
                <div
                  className="Avatar mr-05"
                  style={
                    Object {
                      "height": 20,
                      "width": 20,
                    }
                  }
                >
                  <img
                    src="test"
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "var(--bf-green)",
                        "borderRadius": "50%",
                        "display": "flex",
                        "height": "100%",
                        "justifyContent": "center",
                        "width": "100%",
                      }
                    }
                  />
                </div>
              </Avatar>
              <Component>
                <span
                  className=""
                  data-qa="text"
                  style={
                    Object {
                      "color": "hsl(0, 2%, 22%)",
                      "fontFamily": "Inter",
                      "fontSize": 12,
                      "fontWeight": 400,
                      "textAlign": "initial",
                      "textTransform": "initial",
                    }
                  }
                >
                  Test Guy
                </span>
              </Component>
            </div>
          </span>
        </Component>
      </div>
    </div>
  </Chip>
</UserChip>
`;
